katello/hooks/pre/30-upgrade.rb

```diff
@@ -37,6 +37,43 @@ def migrate_foreman
   Kafo::Helpers.execute('foreman-rake db:migrate')
 end
 
+def postgresql_10_upgrade
+  return unless facts[:os][:release][:major] == '7'
+
+  #Ensure there is at least 1x /var/lib/pgsql free disk space available on the main filesystem
+  current_postgres_dir = '/var/lib/pgsql'
+  new_postgres_dir = '/var/opt/rh/rh-postgresql10/lib/pgsql'
+
+  begin
+    unless File.exist?(new_postgres_dir)
+      new_postgres_dir = nil
+    end
+
+    postgres_size = `du -k -s #{current_postgres_dir}`.split[0].to_i
+
+    if available_space(new_postgres_dir) < postgres_size
+      fail_and_exit 'The postgres upgrade requires at least #{postgres_size} KB of storage.'
+    end
+  rescue StandardError
+    fail_and_exit 'Failed to verify available disk space'
+  end
+
+  stop_services
+  Kafo::Helpers.execute('yum -y install rh-postgresql10-postgresql-server')
+  Kafo::Helpers.execute('scl enable rh-postgresql10 -- postgresql-setup --upgrade')
+  Kafo::Helpers.execute('yum -y remove postgresql postgresql-server')
+  Kafo::Helpers.execute('rm -f /etc/systemd/system/postgresql.service')
+  Kafo::Helpers.execute('yum -y install rh-postgresql10-syspaths')
+end
+
+def available_space(directory = nil)
+  `df -k --total --exclude-type=tmpfs --output=avail #{directory}`.split("\n")[-1].to_i
```

Comment:
`--exclude-type=tmpfs` is technically not really effective because you ask it for a specific directory. Same for `--total`.

Reply:
Alternative implementation that doesn't require shelling out:
```ruby
def available_bytes(directory)
  mountpoints = facts[:mountpoints]
  until (mountpoint = mountpoints[directory])
    directory = File.dirname(directory)
  end
  mountpoint[:available_bytes]
end
```

Reply:
Using this requires passing `available_bytes(:/)` if the directory does not exist? Just understanding how this works for the default case:

e.g
```
{:/=>{:available=>"34.04 GiB", :available_bytes=>36553474048, :capacity=>"13.26%", :device=>"/dev/vda1", :filesystem=>"ext4", :options=>["rw", "seclabel", "relatime", "data=ordered"], :size=>"39.25 GiB", :size_bytes=>42140401664, :used=>"5.20 GiB", :used_bytes=>5586927616}, :"/dev"=>{:available=>"3.81 GiB", :available_bytes=>4092948480, :capacity=>"0%", :device=>"devtmpfs", :filesystem=>"devtmpfs", :options=>["rw", "seclabel", "nosuid", "size=3997020k", "nr_inodes=999255", "mode=755"], :size=>"3.81 GiB", :size_bytes=>4092948480, :used=>"0 bytes", :used_bytes=>0}, :"/dev/hugepages"=>{:available=>"0 bytes", :available_bytes=>0, :capacity=>"100%", :device=>"hugetlbfs", :filesystem=>"hugetlbfs", :options=>["rw", "seclabel", "relatime"], :size=>"0 bytes", :size_bytes=>0, :used=>"0 bytes", :used_bytes=>0}, :"/dev/mqueue"=>{:available=>"0 bytes", :available_bytes=>0, :capacity=>"100%", :device=>"mqueue", :filesystem=>"mqueue", :options=>["rw", "seclabel", "relatime"], :size=>"0 bytes", :size_bytes=>0, :used=>"0 bytes", :used_bytes=>0}, :"/dev/pts"=>{:available=>"0 bytes", :available_bytes=>0, :capacity=>"100%", :device=>"devpts", :filesystem=>"devpts", :options=>["rw", "seclabel", "nosuid", "noexec", "relatime", "gid=5", "mode=620", "ptmxmode=000"], :size=>"0 bytes", :size_bytes=>0, :used=>"0 bytes", :used_bytes=>0}, :"/dev/shm"=>{:available=>"3.82 GiB", :available_bytes=>4100939776, :capacity=>"0.00%", :device=>"tmpfs", :filesystem=>"tmpfs", :options=>["rw", "seclabel", "nosuid", "nodev"], :size=>"3.82 GiB", :size_bytes=>4100960256, :used=>"20.00 KiB", :used_bytes=>20480}, :"/run"=>{:available=>"3.81 GiB", :available_bytes=>4091801600, :capacity=>"0.22%", :device=>"tmpfs", :filesystem=>"tmpfs", :options=>["rw", "seclabel", "nosuid", "nodev", "mode=755"], :size=>"3.82 GiB", :size_bytes=>4100960256, :used=>"8.73 MiB", :used_bytes=>9158656}, :"/run/user/1000"=>{:available=>"782.20 MiB", :available_bytes=>820195328, :capacity=>"0%", :device=>"tmpfs", :filesystem=>"tmpfs", :options=>["rw", "seclabel", "nosuid", "nodev", "relatime", "size=800972k", "mode=700", "uid=1000", "gid=1000"], :size=>"782.20 MiB", :size_bytes=>820195328, :used=>"0 bytes", :used_bytes=>0}, :"/sys/fs/cgroup"=>{:available=>"3.82 GiB", :available_bytes=>4100960256, :capacity=>"0%", :device=>"tmpfs", :filesystem=>"tmpfs", :options=>["ro", "seclabel", "nosuid", "nodev", "noexec", "mode=755"], :size=>"3.82 GiB", :size_bytes=>4100960256, :used=>"0 bytes", :used_bytes=>0}, :"/var/lib/nfs/rpc_pipefs"=>{:available=>"0 bytes", :available_bytes=>0, :capacity=>"100%", :device=>"sunrpc", :filesystem=>"rpc_pipefs", :options=>["rw", "relatime"], :size=>"0 bytes", :size_bytes=>0, :used=>"0 bytes", :used_bytes=>0}}
```

Reply:
It may need a `to_sym`. The idea is that you go "up" until it actually is found. By doing so, you're guaranteed to find the correct mountpoint that `mkdir -p` would also use. A minimal reproducer with debugging:
```ruby
require 'json'

# Taken from lib/kafo/base_context.rb
def symbolize(data)
  case data
  when Hash
    Hash[data.map { |key, value| [key.to_sym, symbolize(value)] }]
  when Array
    data.map { |v| symbolize(v) }
  else
    data
  end
end

FACTS = symbolize(JSON.parse(`facter --json`))

# Actual code

def available_bytes(directory)
  mountpoints = FACTS[:mountpoints]
  puts "Attempting #{directory}"
  until (mountpoint = mountpoints[directory.to_sym])
    raise Exception, "Could not find #{directory}" if directory == '/'
    directory = File.dirname(directory)
    puts "Attempting #{directory}"
  end
  puts "Found #{directory}"
  mountpoint[:available_bytes]
end

puts "Mountpoints: #{FACTS[:mountpoints].keys.inspect}"
puts available_bytes('/tmp/does/not/exist')
```

Output:
```
$ ruby disk_space.rb 
Mountpoints: [:/, :"/boot", :"/boot/efi", :"/dev", :"/dev/hugepages", :"/dev/mqueue", :"/dev/pts", :"/dev/shm", :"/run", :"/run/user/1000", :"/run/user/1000/gvfs", :"/tmp", :"/var/lib/nfs/rpc_pipefs"]
Attempting /tmp/does/not/exist
Attempting /tmp/does/not
Attempting /tmp/does
Attempting /tmp
Found /tmp
8328683520
```